src/ParseServer.js

```diff
@@ -183,6 +183,8 @@ class ParseServer {
     api.use(bodyParser.json({ type: '*/*', limit: maxUploadSize }));
     api.use(middlewares.allowMethodOverride);
     api.use(middlewares.handleParseHeaders);
+    middlewares.handleRateLimit(api, rateLimit);
```

Comment:
Can we put `middlewares.handleRateLimit(api, rateLimit);` before `api.use(middlewares.handleParseHeaders);`, or even earlier?

`middlewares.handleParseHeaders` handles authentication, so we should generally put it before that. Also, I assume the earlier we can add the rate limiter, the less resources an attack consumes on the server side.

Reply:
The only reason I had it that way was because `handleParseHeaders` works out whether the request was with a masterKey or not, and the rate limiter can skip with a masterKey. I moved the session lookup out of `handleParseHeaders` so it should resolve very quickly and use very few resources

Reply:
Got it